### PR TITLE
Fix token encoding crash when compiling without version bubble

### PR DESCRIPTION
**Pull request by fadimounir**
Implement the ability to get token values for typeRefs and memberRefs that are not part of the version bubble. This fixes the ability to compile the CoreFX set without the --large-bubble switch

### CODE REVIEW

**Comment by fadimounir** [comment]
@dotnet/crossgen-contrib

File: src/coreclr/src/tools/crossgen2/ILCompiler.ReadyToRun/Compiler/ReadyToRunSingleAssemblyCompilationModuleGroup.cs
Comment (janvorli) [inline src/coreclr/src/tools/crossgen2/ILCompiler.ReadyToRun/Compiler/ReadyToRunSingleAssemblyCompilationModuleGroup.cs]: Maybe we could use TryAdd instead of ContainsKey/Add - seems like it would do half the amount of lookup.

File: src/coreclr/src/tools/crossgen2/ILCompiler.ReadyToRun/Compiler/SingleMethodCompilationModuleGroup.cs
Comment (janvorli) [inline src/coreclr/src/tools/crossgen2/ILCompiler.ReadyToRun/Compiler/SingleMethodCompilationModuleGroup.cs]: What is the SingleMethodCompilationModuleGroup.cs  good for?

**Comment by jkotas** [comment]
Roud-tripping typeref tokens is ok.

memberRef tokens do not round-trip reliably with version resiliency in the picture. I do not believe that we are doing similar roundtripping in crossgen for this reason (correct me if I am wrong). Can we do the same thing as crossgen here?

File: src/coreclr/src/tools/crossgen2/ILCompiler.ReadyToRun/Compiler/SingleMethodCompilationModuleGroup.cs
Comment (nattress) [reply src/coreclr/src/tools/crossgen2/ILCompiler.ReadyToRun/Compiler/SingleMethodCompilationModuleGroup.cs]: This is a debug aid for devs. If you have a compiler bug and it crashes compiling method X, you can pass a command line `--singlemethodname X` and it'll just compile the thing you care about.

**Comment by davidwrighton** [comment]
@jkotas Yes, round-tripping of generic member-refs doesn't work. (But round-tripping of non-generics is fine).

This code just feels wrong. I agree that we should be just attempting to preserve the token as is done in normal crossgen.

**Comment by fadimounir** [comment]
These changes should only be handling non-generics, so should be ok. But I will look at what crossgen does and compare.

David, why does the code feel wrong?

**Comment by davidwrighton** [comment]
Why are we handling generics any different than non-generics? The critical detail with tokens is that we can't generate new metadata currently in crossgen2. In crossgen1 we handle this by always re-using exactly the same tokens for each fixup as were used in the il stream, but here we seem to be attempting to recreate tokens. Thats what feels odd.

**Comment by fadimounir** [comment]
> but here we seem to be attempting to recreate tokens. Thats what feels odd.

It shouldn’t recreate. It uses the same tokens found in the input assembly (loaded from the typeref and memberref tables in the metadata from the input assembly). Or am I missing something?

Anyways, I agree that I should make some changes here and use the tokens we resolve from the input.

**Comment by fadimounir** [comment]
We're doing the right thing already for methods/fields. It was just about typerefs. There are some cases where typerefs might not necessarily show up in `CORINFO_RESOLVED_TOKEN` structures (ex: `classMustBeLoadedBeforeCodeIsRun`), and these cases were failing during signature writing because the typeref token recording logic we have only works on calls to `resolveToken`.

File: src/coreclr/src/tools/crossgen2/ILCompiler.ReadyToRun/Compiler/DependencyAnalysis/ReadyToRun/ModuleTokenResolver.cs
Comment (jkotas) [inline src/coreclr/src/tools/crossgen2/ILCompiler.ReadyToRun/Compiler/DependencyAnalysis/ReadyToRun/ModuleTokenResolver.cs]: Is this guaranteed to be called from a single thread only?

Comment (fadimounir) [reply src/coreclr/src/tools/crossgen2/ILCompiler.ReadyToRun/Compiler/DependencyAnalysis/ReadyToRun/ModuleTokenResolver.cs]: The signature emission happens in a single thread at the end (otherwise we can have determinism bugs).

File: src/coreclr/src/tools/crossgen2/ILCompiler.ReadyToRun/Compiler/SingleMethodCompilationModuleGroup.cs
Comment (davidwrighton) [reply src/coreclr/src/tools/crossgen2/ILCompiler.ReadyToRun/Compiler/SingleMethodCompilationModuleGroup.cs]: Yes, please implement this. Being able to compile a single method and run it through the compiler all the way to generating a PE file is quite useful.

---
In reply to: [357427440](https://github.com/dotnet/runtime/pull/816#discussion_r357427440) [](ancestors = 357427440)

File: src/coreclr/src/tools/crossgen2/ILCompiler.ReadyToRun/Compiler/DependencyAnalysis/ReadyToRun/ModuleTokenResolver.cs
Comment (fadimounir) [reply src/coreclr/src/tools/crossgen2/ILCompiler.ReadyToRun/Compiler/DependencyAnalysis/ReadyToRun/ModuleTokenResolver.cs]: Verified under debugger too.

File: src/coreclr/src/tools/crossgen2/ILCompiler.ReadyToRun/Compiler/SingleMethodCompilationModuleGroup.cs
Comment (fadimounir) [reply src/coreclr/src/tools/crossgen2/ILCompiler.ReadyToRun/Compiler/SingleMethodCompilationModuleGroup.cs]: Done

**Comment by fadimounir** [comment]
Ping for code review :) ... CI failures are infra failures.

**Review by trylek — APPROVED** [review]
To me this looks good. While I don't see any hard requirement why we need to manually scan typerefs, in accordance with JanK's opinion I believe that should be safe and ultimately let us simplify some of the code in ModuleTokenResolver (basically deleting all the type token harvesting code). @davidwrighton expressed some concerns in the PR discussion though so I would personally prefer to wait for his approval before merging this change in.

File: src/coreclr/src/tools/crossgen2/ILCompiler.ReadyToRun/Compiler/ReadyToRunSingleAssemblyCompilationModuleGroup.cs
Comment (davidwrighton) [inline src/coreclr/src/tools/crossgen2/ILCompiler.ReadyToRun/Compiler/ReadyToRunSingleAssemblyCompilationModuleGroup.cs]: It feels like you need a try/catch here in case the typeref isn't resolvable.

File: src/coreclr/src/tools/crossgen2/ILCompiler.ReadyToRun/Compiler/SingleMethodCompilationModuleGroup.cs
Comment (davidwrighton) [inline src/coreclr/src/tools/crossgen2/ILCompiler.ReadyToRun/Compiler/SingleMethodCompilationModuleGroup.cs]: Also needs protection from bad typerefs

**Review by davidwrighton — CHANGES_REQUESTED** [review]
Looks good, but you'll also need to address dangling type refs

File: src/coreclr/src/tools/crossgen2/ILCompiler.ReadyToRun/Compiler/ReadyToRunSingleAssemblyCompilationModuleGroup.cs
Comment (fadimounir) [reply src/coreclr/src/tools/crossgen2/ILCompiler.ReadyToRun/Compiler/ReadyToRunSingleAssemblyCompilationModuleGroup.cs]: Fixed

File: src/coreclr/src/tools/crossgen2/ILCompiler.ReadyToRun/Compiler/SingleMethodCompilationModuleGroup.cs
Comment (fadimounir) [reply src/coreclr/src/tools/crossgen2/ILCompiler.ReadyToRun/Compiler/SingleMethodCompilationModuleGroup.cs]: Fixed

**Review by davidwrighton — APPROVED** [review]

**Pull request merged**
